### ISSUE TITLE
Pytorch BatchNorm2D Unstable

### ISSUE THREAD

**Issue by todpole3** [pytorch/pytorch#12812]
I was tuning the following customed nn module:

```class ConvE(nn.Module):
    def __init__(self, args, num_entities):
        super(ConvE, self).__init__()
        self.entity_dim = args.entity_dim
        self.relation_dim = args.relation_dim
        assert(args.emb_2D_d1 * args.emb_2D_d2 == args.entity_dim)
        assert(args.emb_2D_d1 * args.emb_2D_d2 == args.relation_dim)
        self.emb_2D_d1 = args.emb_2D_d1
        self.emb_2D_d2 = args.emb_2D_d2
        self.num_out_channels = args.num_out_channels
        self.w_d = args.kernel_size
        self.HiddenDropout = nn.Dropout(args.hidden_dropout_rate)
        self.FeatureDropout = nn.Dropout(args.feat_dropout_rate)

        # stride = 1, padding = 0, dilation = 1, groups = 1
        self.conv1 = nn.Conv2d(1, self.num_out_channels, (self.w_d, self.w_d), 1, 0)
        self.bn0 = nn.BatchNorm2d(1)
        self.bn1 = nn.BatchNorm2d(self.num_out_channels)
        self.bn2 = nn.BatchNorm1d(self.entity_dim)
        self.register_parameter('b', nn.Parameter(torch.zeros(num_entities)))
        h_out = 2 * self.emb_2D_d1 - self.w_d + 1
        w_out = self.emb_2D_d2 - self.w_d + 1
        self.feat_dim = self.num_out_channels * h_out * w_out
        self.fc = nn.Linear(self.feat_dim, self.entity_dim)

    def forward(self, e1, r, kg):
        E1 = kg.get_entity_embeddings(e1).view(-1, 1, self.emb_2D_d1, self.emb_2D_d2)
        R = kg.get_relation_embeddings(r).view(-1, 1, self.emb_2D_d1, self.emb_2D_d2)
        E2 = kg.get_all_entity_embeddings()

        stacked_inputs = torch.cat([E1, R], 2)
        stacked_inputs = self.bn0(stacked_inputs)

        X = self.conv1(stacked_inputs)
        X = self.bn1(X)                            # <= problem line
        X = F.relu(X)
        X = self.FeatureDropout(X)
        X = X.view(-1, self.feat_dim)
        X = self.fc(X)
        X = self.HiddenDropout(X)
        X = self.bn2(X)
        X = F.relu(X)
        X = torch.mm(X, E2.transpose(1, 0))
        X += self.b.expand_as(X)

        S = F.sigmoid(X)
        return S```

If I include the "problem line" in code, the model performance became very unstable and the loss declines slowly. If I commented out the line, the model performance becomes normal.

The strange part is that it used to be fine to include this line in my code and this problem caught my attention after I switched to 0.4.0+.

I wonder if it is caused by any changed in the BatchNorm2d operator.

Below the pink curve is the result obtained w/ the problem line and the orange curve is obtained w/o it.
![screen shot 2018-10-18 at 1 32 19 am](https://user-images.githubusercontent.com/4227871/47141498-af5e4e80-d275-11e8-8a90-356829ce8768.png)

The experiment code repo can be found here:
https://github.com/salesforce/MultiHopKG/blob/master/src/emb/fact_network.py#L131

**Comment by SsnL** [comment]
Hi! Thanks for the report. There hasn't been changes to BN2d as far as I can remember. Could you try to come up with a small, runnable script that reproduce the issue? Thank you!
